### PR TITLE
Scaffold IntegrationTest framework to support recording and slides tests

### PR DESCRIPTION
**Pull request by reddvl1980**
- fix FakeVideoSource to work with bitrates where frame size > MTU;
- add FakeVideoDecoder to facilitate per-frame checks.

### CODE REVIEW

File: test/bwe/FakeVideoSource.cpp [hunk 1]
```diff
@@ -21,27 +25,44 @@ FakeVideoSource::FakeVideoSource(memory::PacketPoolAllocator& allocator, uint32_
...
+    FakeVideoFrameData frameData;
+    frameData.frameNum = _counter;
+    frameData.keyFrame = keyFrame;
+    frameData.lastPacketInFrame = lastInFrame;
+    frameData.packetId = _packetsInFrame;
+    frameData.ssrc = _ssrc;
+    frameData.endpointIdHash = _endpointIdHash;
+    frameData.tag = _tag;
+
+    memcpy(payload + VP8_HEADER_SIZE, &frameData, sizeof(frameData));
```
Comment (olofkallander) [inline test/bwe/FakeVideoSource.cpp:65]: std::memcpy

File: test/integration/emulator/FakeVideoDecoder.h [hunk 2]
```diff
@@ -0,0 +1,244 @@
...
+
+#if DEBUG_FAKE_VIDEO_DECODER
+#define FAKE_VIDEO_DECODER_LOG(fmt, ...) logger::debug(fmt, ##__VA_ARGS__)
+#else
+#define FAKE_VIDEO_DECODER_LOG(fmt, ...)
+#endif
+
+namespace emulator
+{
+class FakeVideoDecoder
```
Comment (olofkallander) [inline test/integration/emulator/FakeVideoDecoder.h:17]: I think this large class deserves its own cpp file

File: test/integration/emulator/FakeVideoDecoder.h [hunk 3]
```diff
@@ -0,0 +1,244 @@
...
+
+public:
+    FakeVideoDecoder(const size_t endpointIdHash, const size_t instanceId)
+        : _lastDecodedFrameNum(0),
+          _ssrc(0),
+          _lastDecodedFrameTs(0),
+          _loggableId((std::string("FakeVideoDecoder-") + std::to_string(endpointIdHash)).c_str(), instanceId)
+    {
+    }
+    void process(const uint8_t* packet, uint32_t length, const uint64_t timestamp)
```
Comment (olofkallander) [inline test/integration/emulator/FakeVideoDecoder.h:65]: new line before

File: test/integration/emulator/FakeVideoDecoder.h [hunk 4]
```diff
@@ -0,0 +1,244 @@
...
+        FAKE_VIDEO_DECODER_LOG("Reset packet cache (size = %zu)", _loggableId.c_str(), _packets.size());
+        _packets.clear();
+    }
+
+    Stats getStats() const { return _stats; }
+
+private:
+    void decodeAssembledFrames(const uint64_t timestamp)
+    {
+        while (_assembledFrames.size())
```
Comment (olofkallander) [inline test/integration/emulator/FakeVideoDecoder.h:116]: while(!_assembledFrames.empty())

File: test/integration/emulator/FakeVideoDecoder.h [hunk 5]
```diff
@@ -0,0 +1,244 @@
...
+            [&data](const fakenet::FakeVideoFrameData& packet) {
+                return packet.frameNum == data.frameNum && packet.ssrc == data.ssrc;
+            });
+
+        framePackets.sort([](const fakenet::FakeVideoFrameData& one, const fakenet::FakeVideoFrameData& another) {
+            return one.packetId < another.packetId;
+        });
+
+        const auto numPackets = framePackets.size();
+        const bool result =
```
Comment (olofkallander) [inline test/integration/emulator/FakeVideoDecoder.h:180]: This means something other than "result". It is true when the frame is complete and contains all packets in the frame, right?

File: test/integration/emulator/SfuClient.h [hunk 6]
```diff
@@ -576,6 +606,12 @@ class SfuClient : public transport::DataReceiver
                     }
                 }
             }
+
+            if (rtpHeader->padding == 1)
```
Comment (olofkallander) [inline test/integration/emulator/SfuClient.h:610]: Should we check that it is RTX payload type also. Padding is strictly allowed and ignored in real vp8 packets.

File: test/integration/BarbellTest.cpp [hunk 7]
```diff
@@ -252,23 +227,34 @@ TEST_F(BarbellTest, packetLossViaBarbell)
...
+
+            logger::debug("%s received video pkts %" PRIu64, "bbTest", clientName.c_str(), videoReceiveStats.packets);
+            logVideoReceive(clientName.c_str(), *group.clients[id]);
+            logTransportSummary(clientName.c_str(), group.clients[id]->_transport.get(), transportSummary);
+
+            auto allStreamsVideoStats = group.clients[id]->getActiveVideoDecoderStats();
+            EXPECT_EQ(allStreamsVideoStats.size(), 2);
+            for (const auto& videoStats : allStreamsVideoStats)
+            {
+                double fps = (double)utils::Time::sec / (double)videoStats.averageFrameRateDelta;
```
Comment (olofkallander) [inline test/integration/BarbellTest.cpp:246]: const

File: test/integration/BarbellTest.cpp [hunk 8]
```diff
@@ -252,23 +227,34 @@ TEST_F(BarbellTest, packetLossViaBarbell)
...
-            videoReceiveStats.lostPackets);
-        logTransportSummary("client2", group.clients[1]->_transport.get(), transportSummary1);
-        logTransportSummary("client3", group.clients[2]->_transport.get(), transportSummary2);
-        EXPECT_GE(group.clients[0]->getVideoPacketsReceived(),
-            transportSummary1.begin()->second.packetsSent + transportSummary2.begin()->second.packetsSent - 100);
-        EXPECT_NEAR(group.clients[0]->getVideoPacketsReceived(),
-            transportSummary1.begin()->second.packetsSent + transportSummary2.begin()->second.packetsSent,
-            200);
+            std::unordered_map<uint32_t, transport::ReportSummary> transportSummary;
+            auto videoReceiveStats = group.clients[id]->_transport->getCumulativeVideoReceiveCounters();
```
Comment (olofkallander) [inline test/integration/BarbellTest.cpp:235]: const

File: test/integration/BarbellTest.cpp [hunk 9]
```diff
@@ -390,12 +382,18 @@ TEST_F(BarbellTest, simpleBarbell)
...
         logTransportSummary("client3", group.clients[2]->_transport.get(), transportSummary2);
 
-        EXPECT_NEAR(videoReceiveStats.packets,
-            transportSummary1.begin()->second.packetsSent + transportSummary2.begin()->second.packetsSent,
-            25);
+        auto allStreamsVideoStats = group.clients[0]->getActiveVideoDecoderStats();
+        EXPECT_EQ(allStreamsVideoStats.size(), 2);
+        for (const auto& videoStats : allStreamsVideoStats)
+        {
+            double fps = (double)utils::Time::sec / (double)videoStats.averageFrameRateDelta;
```
Comment (olofkallander) [inline test/integration/BarbellTest.cpp:393]: const

File: test/integration/IntegrationTest.h [hunk 10]
```diff
@@ -15,6 +16,78 @@
...
+    }
+    logger::info("%s\n", "VideoDecoderInfo", logLine.str().c_str());
+}
+
+template <typename T>
+void logTransportSummary(const char* clientName, transport::RtcTransport* transport, T& summary)
+{
+    for (auto& report : summary)
+    {
+        auto bitrate = report.second.rtpFrequency * report.second.octets /
```
Comment (olofkallander) [inline test/integration/IntegrationTest.h:77]: const

File: test/integration/emulator/SfuClient.h (hunk 6 above)
Comment (reddvl1980) [reply test/integration/emulator/SfuClient.h:610]: Addressed.

File: test/bwe/FakeVideoSource.cpp (hunk 1 above)
Comment (reddvl1980) [reply test/bwe/FakeVideoSource.cpp:65]: Fixed.

File: test/integration/emulator/FakeVideoDecoder.h (hunk 5 above)
Comment (reddvl1980) [reply test/integration/emulator/FakeVideoDecoder.h:180]: Agree.

File: test/integration/emulator/FakeVideoDecoder.h (hunk 2 above)
Comment (reddvl1980) [reply test/integration/emulator/FakeVideoDecoder.h:17]: Agree.

File: test/integration/emulator/FakeVideoDecoder.h [hunk 11]
```diff
@@ -0,0 +1,244 @@
...
+
+                updateStats(nextFrame->endpointIdHash, nextFrame->tag, reorderQueueSize, timestamp);
+
+                _assembledFrames.remove_if([nextFrame](const AssembledFrame& frame) {
+                    return frame.ssrc != nextFrame->ssrc || frame.frameNum <= nextFrame->frameNum;
+                });
+            }
+            else
+            {
+                break;
```
Comment (RicardoMDomingues) [inline test/integration/emulator/FakeVideoDecoder.h:147]: [minor] Please move break to close to the main condition, I mean inverse the condition on If-statement and break there. So it easier to understand the precondition fot the rest to run.
You will not need an "else" either

File: transport/RtpSenderState.h [hunk 12]
```diff
@@ -34,6 +34,8 @@ struct ReportSummary
     uint32_t packetsSent = 0;
     uint32_t initialRtpTimestamp = 0;
     uint32_t rtpTimestamp = 0;
+    uint64_t octets = 0;
+    uint32_t rtpFrequency;
```
Comment (RicardoMDomingues) [inline transport/RtpSenderState.h:38]: perhaps it should be initialised like the other members

**Review by RicardoMDomingues — APPROVED** [review]
LGTM. Minor comments only

**Comment by olofkallander** [comment]
LGTM

**Review by olofkallander — APPROVED** [review]

**Pull request merged**